src/main/kotlin/uk/gov/justice/digital/hmpps/sendlegalmailtoprisonsapi/barcode/BarcodeService.kt

```diff
@@ -36,15 +47,34 @@ class BarcodeService(
       .also { barcode ->
         with(barcodeEventService) {
           createEvent(barcode, userId, BarcodeEventType.CHECKED, location, sourceIp)
-          checkForCreated(barcode)
+          try {
+            checkForCreated(barcode)
+          } catch (e: ResourceNotFoundException) {
+            // if ResourceNotFoundException thrown track the event
+            // else track the event post call to lookupRecipient
+            trackScanEvent(ScanEventType.NON_EXISTENT_BARCODE, barcode, userId, location)
+            throw e
+          }
 
           barcodeRecipientService.getBarcodeRecipient(barcode)
             ?.lookupRecipient()
             ?: log.info { "No BarcodeRecipient record for barcode ${barcode.code}" }
 
-          checkForDuplicate(barcode, userId, location, sourceIp)
-          checkForExpired(barcode, userId, location, sourceIp)
-          checkForRandomSecurityCheck(barcode, userId, location, sourceIp)
+          try {
+            checkForDuplicate(barcode, userId, location, sourceIp)
+            checkForExpired(barcode, userId, location, sourceIp)
+            checkForRandomSecurityCheck(barcode, userId, location, sourceIp)
+
+            // if no validation exception thrown above it means that the barcode is ready for delivery
+            trackScanEvent(ScanEventType.READY_FOR_DELIVERY, barcode, userId, location)
+          } catch (e: ValidationException) {
+            when (e.errorCode.code) {
+              "DUPLICATE" -> trackScanEvent(ScanEventType.DUPLICATE, barcode, userId, location)
+              "EXPIRED" -> trackScanEvent(ScanEventType.EXPIRED, barcode, userId, location)
+              "RANDOM_CHECK" -> trackScanEvent(ScanEventType.RANDOM_CHECK, barcode, userId, location)
+            }
+            throw e
```

Comment:
`
trackScanEvent(ScanEventType.valueOf(e.errorCode.code), barcode, userId, location)
`
Would be better I think

Reply:
Changed it initially but then decided to keep it the same as there are a few ValidationExceptions thrown and this is the safest code possible.

Reply:
updated to a single when clause as suggested.